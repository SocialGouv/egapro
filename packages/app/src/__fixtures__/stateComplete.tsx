import {
  CategorieSocioPro,
  PeriodeDeclaration,
  Structure,
  TranchesAges,
  TrancheEffectifs
} from "../globals.d";
import AppReducer from "../AppReducer";

const actionInitiateState = {
  type: "initiateState" as "initiateState",
  data: {}
};

const actionUpdateInformationsSimulation = {
  type: "updateInformationsSimulation" as "updateInformationsSimulation",
  data: {
    nomEntreprise: "BigCorp",
    trancheEffectifs: "1000 et plus" as TrancheEffectifs,
    debutPeriodeReference: "2019-01-01",
    finPeriodeReference: "2019-12-31"
  }
};

const actionUpdateEffectif = {
  type: "updateEffectif" as "updateEffectif",
  data: {
    nombreSalaries: [
      {
        categorieSocioPro: CategorieSocioPro.Ouvriers,
        tranchesAges: [
          {
            trancheAge: TranchesAges.MoinsDe30ans,
            nombreSalariesFemmes: 23,
            nombreSalariesHommes: 32
          },
          {
            trancheAge: TranchesAges.De30a39ans,
            nombreSalariesFemmes: 12,
            nombreSalariesHommes: 13
          },
          {
            trancheAge: TranchesAges.De40a49ans,
            nombreSalariesFemmes: 34,
            nombreSalariesHommes: 7
          },
          {
            trancheAge: TranchesAges.PlusDe50ans,
            nombreSalariesFemmes: 5,
            nombreSalariesHommes: 21
          }
        ]
      },
      {
        categorieSocioPro: CategorieSocioPro.Employes,
        tranchesAges: [
          {
            trancheAge: TranchesAges.MoinsDe30ans,
            nombreSalariesFemmes: 63,
            nombreSalariesHommes: 25
          },
          {
            trancheAge: TranchesAges.De30a39ans,
            nombreSalariesFemmes: 52,
            nombreSalariesHommes: 62
          },
          {
            trancheAge: TranchesAges.De40a49ans,
            nombreSalariesFemmes: 16,
            nombreSalariesHommes: 18
          },
          {
            trancheAge: TranchesAges.PlusDe50ans,
            nombreSalariesFemmes: 27,
            nombreSalariesHommes: 19
          }
        ]
      },
      {
        categorieSocioPro: CategorieSocioPro.Techniciens,
        tranchesAges: [
          {
            trancheAge: TranchesAges.MoinsDe30ans,
            nombreSalariesFemmes: 14,
            nombreSalariesHommes: 15
          },
          {
            trancheAge: TranchesAges.De30a39ans,
            nombreSalariesFemmes: 4,
            nombreSalariesHommes: 5
          },
          {
            trancheAge: TranchesAges.De40a49ans,
            nombreSalariesFemmes: 6,
            nombreSalariesHommes: 8
          },
          {
            trancheAge: TranchesAges.PlusDe50ans,
            nombreSalariesFemmes: 7,
            nombreSalariesHommes: 7
          }
        ]
      },
      {
        categorieSocioPro: CategorieSocioPro.Cadres,
        tranchesAges: [
          {
            trancheAge: TranchesAges.MoinsDe30ans,
            nombreSalariesFemmes: 7,
            nombreSalariesHommes: 8
          },
          {
            trancheAge: TranchesAges.De30a39ans,
            nombreSalariesFemmes: 10,
            nombreSalariesHommes: 8
          },
          {
            trancheAge: TranchesAges.De40a49ans,
            nombreSalariesFemmes: 13,
            nombreSalariesHommes: 13
          },
          {
            trancheAge: TranchesAges.PlusDe50ans,
            nombreSalariesFemmes: 16,
            nombreSalariesHommes: 19
          }
        ]
      }
    ]
  }
};

const actionUpdateIndicateurUnCsp = {
  type: "updateIndicateurUnCsp" as "updateIndicateurUnCsp",
  data: {
    remunerationAnnuelle: [
      {
        categorieSocioPro: CategorieSocioPro.Ouvriers,
        tranchesAges: [
          {
            trancheAge: TranchesAges.MoinsDe30ans,
            remunerationAnnuelleBrutFemmes: 23000,
            remunerationAnnuelleBrutHommes: 25000
          },
          {
            trancheAge: TranchesAges.De30a39ans,
            remunerationAnnuelleBrutFemmes: 24000,
            remunerationAnnuelleBrutHommes: 26000
          },
          {
            trancheAge: TranchesAges.De40a49ans,
            remunerationAnnuelleBrutFemmes: 25500,
            remunerationAnnuelleBrutHommes: 26000
          },
          {
            trancheAge: TranchesAges.PlusDe50ans,
            remunerationAnnuelleBrutFemmes: 27500,
            remunerationAnnuelleBrutHommes: 28000
          }
        ]
      },
      {
        categorieSocioPro: CategorieSocioPro.Employes,
        tranchesAges: [
          {
            trancheAge: TranchesAges.MoinsDe30ans,
            remunerationAnnuelleBrutFemmes: 23000,
            remunerationAnnuelleBrutHommes: 25000
          },
          {
            trancheAge: TranchesAges.De30a39ans,
            remunerationAnnuelleBrutFemmes: 24000,
            remunerationAnnuelleBrutHommes: 26000
          },
          {
            trancheAge: TranchesAges.De40a49ans,
            remunerationAnnuelleBrutFemmes: 31000,
            remunerationAnnuelleBrutHommes: 33000
          },
          {
            trancheAge: TranchesAges.PlusDe50ans,
            remunerationAnnuelleBrutFemmes: 39000,
            remunerationAnnuelleBrutHommes: 43000
          }
        ]
      },
      {
        categorieSocioPro: CategorieSocioPro.Techniciens,
        tranchesAges: [
          {
            trancheAge: TranchesAges.MoinsDe30ans,
            remunerationAnnuelleBrutFemmes: 26000,
            remunerationAnnuelleBrutHommes: 28000
          },
          {
            trancheAge: TranchesAges.De30a39ans,
            remunerationAnnuelleBrutFemmes: 27000,
            remunerationAnnuelleBrutHommes: 29000
          },
          {
            trancheAge: TranchesAges.De40a49ans,
            remunerationAnnuelleBrutFemmes: 34000,
            remunerationAnnuelleBrutHommes: 36000
          },
          {
            trancheAge: TranchesAges.PlusDe50ans,
            remunerationAnnuelleBrutFemmes: 42000,
            remunerationAnnuelleBrutHommes: 46000
          }
        ]
      },
      {
        categorieSocioPro: CategorieSocioPro.Cadres,
        tranchesAges: [
          {
            trancheAge: TranchesAges.MoinsDe30ans,
            remunerationAnnuelleBrutFemmes: 36000,
            remunerationAnnuelleBrutHommes: 38000
          },
          {
            trancheAge: TranchesAges.De30a39ans,
            remunerationAnnuelleBrutFemmes: 37000,
            remunerationAnnuelleBrutHommes: 39000
          },
          {
            trancheAge: TranchesAges.De40a49ans,
            remunerationAnnuelleBrutFemmes: 44000,
            remunerationAnnuelleBrutHommes: 46000
          },
          {
            trancheAge: TranchesAges.PlusDe50ans,
            remunerationAnnuelleBrutFemmes: 52000,
            remunerationAnnuelleBrutHommes: 56000
          }
        ]
      }
    ]
  }
};

const actionUpdateIndicateurUnCoefName = {
  type: "updateIndicateurUnCoef" as "updateIndicateurUnCoef",
  data: {
    coefficient: [
      {
        name: "Business Developers"
      }
    ]
  }
};

const actionUpdateIndicateurUnCoefNombreSalaries = {
  type: "updateIndicateurUnCoef" as "updateIndicateurUnCoef",
  data: {
    coefficient: [
      {
        tranchesAges: [
          {
            trancheAge: TranchesAges.MoinsDe30ans,
            nombreSalariesFemmes: 5,
            nombreSalariesHommes: 4
          },
          {
            trancheAge: TranchesAges.De30a39ans,
            nombreSalariesFemmes: 2,
            nombreSalariesHommes: 6
          },
          {
            trancheAge: TranchesAges.De40a49ans,
            nombreSalariesFemmes: 5,
            nombreSalariesHommes: 3
          },
          {
            trancheAge: TranchesAges.PlusDe50ans,
            nombreSalariesFemmes: 1,
            nombreSalariesHommes: 3
          }
        ]
      }
    ]
  }
};

const actionUpdateIndicateurUnCoefRemuneration = {
  type: "updateIndicateurUnCoef" as "updateIndicateurUnCoef",
  data: {
    coefficient: [
      {
        tranchesAges: [
          {
            trancheAge: TranchesAges.MoinsDe30ans,
            remunerationAnnuelleBrutFemmes: 25000,
            remunerationAnnuelleBrutHommes: 24000
          },
          {
            trancheAge: TranchesAges.De30a39ans,
            remunerationAnnuelleBrutFemmes: 32000,
            remunerationAnnuelleBrutHommes: 32000
          },
          {
            trancheAge: TranchesAges.De40a49ans,
            remunerationAnnuelleBrutFemmes: 41000,
            remunerationAnnuelleBrutHommes: 43000
          },
          {
            trancheAge: TranchesAges.PlusDe50ans,
            remunerationAnnuelleBrutFemmes: 49000,
            remunerationAnnuelleBrutHommes: 51500
          }
        ]
      }
    ]
  }
};

const actionUpdateIndicateurDeux = {
  type: "updateIndicateurDeux" as "updateIndicateurDeux",
  data: {
    presenceAugmentation: false,
    tauxAugmentation: [
      {
        categorieSocioPro: CategorieSocioPro.Ouvriers,
        tauxAugmentationFemmes: 2.2,
        tauxAugmentationHommes: 3.5
      },
      {
        categorieSocioPro: CategorieSocioPro.Employes,
        tauxAugmentationFemmes: 1.3,
        tauxAugmentationHommes: 2.2
      },
      {
        categorieSocioPro: CategorieSocioPro.Techniciens,
        tauxAugmentationFemmes: 8.02,
        tauxAugmentationHommes: 6.92
      },
      {
        categorieSocioPro: CategorieSocioPro.Cadres,
        tauxAugmentationFemmes: 12.15,
        tauxAugmentationHommes: 13.56
      }
    ]
  }
};

const actionUpdateIndicateurTrois = {
  type: "updateIndicateurTrois" as "updateIndicateurTrois",
  data: {
    presencePromotion: false,
    tauxPromotion: [
      {
        categorieSocioPro: CategorieSocioPro.Ouvriers,
        tauxPromotionFemmes: 1.1,
        tauxPromotionHommes: 2.2
      },
      {
        categorieSocioPro: CategorieSocioPro.Employes,
        tauxPromotionFemmes: 0.3,
        tauxPromotionHommes: 2.0
      },
      {
        categorieSocioPro: CategorieSocioPro.Techniciens,
        tauxPromotionFemmes: 2,
        tauxPromotionHommes: 3
      },
      {
        categorieSocioPro: CategorieSocioPro.Cadres,
        tauxPromotionFemmes: 4,
        tauxPromotionHommes: 5.55
      }
    ]
  }
};

const actionUpdateIndicateurDeuxTrois = {
  type: "updateIndicateurDeuxTrois" as "updateIndicateurDeuxTrois",
  data: {
    presenceAugmentationPromotion: false,
    nombreAugmentationPromotionFemmes: 1,
    nombreAugmentationPromotionHommes: 2,
    periodeDeclaration: "unePeriodeReference" as PeriodeDeclaration
  }
};

const actionUpdateIndicateurQuatre = {
  type: "updateIndicateurQuatre" as "updateIndicateurQuatre",
  data: {
    presenceCongeMat: false,
    nombreSalarieesPeriodeAugmentation: 7,
    nombreSalarieesAugmentees: 7
  }
};

const actionUpdateIndicateurCinq = {
  type: "updateIndicateurCinq" as "updateIndicateurCinq",
  data: {
    nombreSalariesHommes: 6,
    nombreSalariesFemmes: 4
  }
};

const actionUpdateInformationsEntreprise = {
  type: "updateInformationsEntreprise" as "updateInformationsEntreprise",
  data: {
    nomEntreprise: "acme",
    siren: "1234",
    codeNaf: "5678",
    region: "Auvergne-Rhône-Alpes",
    departement: "Drôme",
    adresse: "30 rue des alouettes",
    codePostal: "12345",
    commune: "Trifouilly",
    structure: "Unité Economique et Sociale (UES)" as Structure,
    nomUES: "nom d'une UES",
    nombreEntreprises: 2,
    entreprisesUES: [
      { nom: "entreprise 1", siren: "12345" },
      { nom: "entreprise 2", siren: "67890" }
    ]
  }
};

const actionUpdateInformationsDeclarant = {
  type: "updateInformationsDeclarant" as "updateInformationsDeclarant",
  data: {
    nom: "Daffy",
    prenom: "Duck",
    tel: "0123456789",
    email: "daffy.duck@example.com"
  }
};

const actionUpdateInformationsComplementaires = {
  type: "updateInformationsComplementaires" as "updateInformationsComplementaires",
  data: {
    dateConsultationCSE: "01/02/2019",
    anneeDeclaration: 2020,
    datePublication: "01/02/2020",
    lienPublication: "https://example.com"
  }
};

const actionUpdateDeclaration = {
  type: "updateDeclaration" as "updateDeclaration",
  data: {}
};

// fast pipe, I miss you in JS…
const stateDefault = AppReducer(
  AppReducer(
    AppReducer(
      AppReducer(
        AppReducer(
          AppReducer(
            AppReducer(
              AppReducer(
                AppReducer(
                  AppReducer(
                    AppReducer(
                      AppReducer(
                        AppReducer(
                          AppReducer(
                            AppReducer(
                              AppReducer(undefined, actionInitiateState),
                              actionUpdateInformationsSimulation
                            ),
                            actionUpdateEffectif
                          ),
                          actionUpdateIndicateurUnCsp
                        ),
                        actionUpdateIndicateurUnCoefName
                      ),
                      actionUpdateIndicateurUnCoefNombreSalaries
                    ),
                    actionUpdateIndicateurUnCoefRemuneration
                  ),
                  actionUpdateIndicateurDeux
                ),
                actionUpdateIndicateurTrois
              ),
              actionUpdateIndicateurDeuxTrois
            ),
            actionUpdateIndicateurQuatre
          ),
          actionUpdateIndicateurCinq
        ),
        actionUpdateInformationsEntreprise
      ),
      actionUpdateInformationsDeclarant
    ),
    actionUpdateInformationsComplementaires
  ),
  actionUpdateDeclaration
);

export default stateDefault;
