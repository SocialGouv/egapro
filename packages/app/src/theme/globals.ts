export const globals = {
  styles: {
    global: {
      "html, body, #root": {
        height: "100%",
      },
      body: {
        fontFamily: '"Cabin", -apple-system, sans-serif',
      },
    },
  },
};
