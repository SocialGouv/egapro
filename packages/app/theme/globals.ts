const styles = {
  styles: {
    global: {
      "html, body, #root": {
        height: "100%",
      },
      body: {
        fontFamily: '"Cabin", -apple-system, sans-serif',
      },
    },
  },
}
export default styles
