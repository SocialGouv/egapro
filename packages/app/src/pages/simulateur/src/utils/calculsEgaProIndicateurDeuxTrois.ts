import { AppState } from "../globals"

import calculIndicateurUn from "./calculsEgaProIndicateurUn"
import totalNombreSalaries from "./totalNombreSalaries"

import {
  calculEcartTauxAugmentation,
  calculIndicateurEcartAugmentation,
  calculIndicateurSexeSurRepresente,
  calculIndicateurEcartAugmentationAbsolute,
} from "../utils/calculsEgaProIndicateurDeux"
import { roundDecimal } from "./number"

const barem = [35, 35, 35, 25, 25, 25, 15, 15, 15, 15, 15, 0]

///////////////////////
// INDICATEUR 2 ET 3 //
///////////////////////

// // ETP
export const calculEcartTauxAugmentationPromotion = calculEcartTauxAugmentation

// // IEP
export const calculIndicateurEcartAugmentationPromotion = calculIndicateurEcartAugmentation

export const calculIndicateurEcartAugmentationPromotionAbsolute = calculIndicateurEcartAugmentationAbsolute

// // IC
export const calculEffectifsIndicateurCalculable = (
  totalNombreSalariesHommes: number | undefined,
  totalNombreSalariesFemmes: number | undefined,
): boolean => {
  return (
    totalNombreSalariesHommes !== undefined &&
    totalNombreSalariesFemmes !== undefined &&
    totalNombreSalariesHommes >= 5 &&
    totalNombreSalariesFemmes >= 5
  )
}

export const calculIndicateurCalculable = (
  presenceAugmentationPromotion: boolean,
  effectifsIndicateurCalculable: boolean,
): boolean => {
  return presenceAugmentationPromotion && effectifsIndicateurCalculable
}

export const calculTaux = (
  nombreSalaries: number | undefined,
  totalNombreSalaries: number | undefined,
): number | undefined =>
  nombreSalaries !== undefined && totalNombreSalaries !== undefined && totalNombreSalaries > 0
    ? nombreSalaries / totalNombreSalaries
    : undefined

export const calculPlusPetitNombreSalaries = (
  totalNombreSalariesHommes: number | undefined,
  totalNombreSalariesFemmes: number | undefined,
): "hommes" | "femmes" | undefined => {
  if (
    totalNombreSalariesFemmes === totalNombreSalariesHommes ||
    totalNombreSalariesFemmes === undefined ||
    totalNombreSalariesHommes === undefined
  ) {
    return undefined
  }
  return totalNombreSalariesHommes > totalNombreSalariesFemmes ? "femmes" : "hommes"
}

export const calculIndicateurEcartNombreEquivalentSalaries = (
  indicateurEcartAugmentationPromotionAbsolute: number | undefined,
  totalNombreSalariesHommes: number | undefined,
  totalNombreSalariesFemmes: number | undefined,
): number | undefined => {
  return indicateurEcartAugmentationPromotionAbsolute !== undefined &&
    totalNombreSalariesHommes !== undefined &&
    totalNombreSalariesFemmes !== undefined
    ? roundDecimal(
        (indicateurEcartAugmentationPromotionAbsolute *
          Math.min(totalNombreSalariesHommes, totalNombreSalariesFemmes)) /
          100,
        1,
      )
    : undefined
}

// // NOTE

export const calculBarem = (indicateur: number): number => {
  return barem[Math.min(barem.length - 1, Math.ceil(Math.max(0, roundDecimal(indicateur, 1))))]
}

export const calculNote = (
  ecartTaux: number | undefined,
  ecartNombreSalaries: number | undefined,
  noteIndicateurUn: number | undefined,
  indicateurUnSexeSurRepresente: "hommes" | "femmes" | undefined,
  indicateurDeuxTroisSexeSurRepresente: "hommes" | "femmes" | undefined,
): {
  note: number | undefined
  correctionMeasure: boolean
  noteEcartTaux: number | undefined
  noteEcartNombreSalaries: number | undefined
} => {
  const noteEcartTaux = ecartTaux !== undefined ? calculBarem(ecartTaux) : undefined
  const noteEcartNombreSalaries = ecartNombreSalaries !== undefined ? calculBarem(ecartNombreSalaries) : undefined
  if (
    noteIndicateurUn !== undefined &&
    noteIndicateurUn < 40 &&
    indicateurUnSexeSurRepresente &&
    indicateurDeuxTroisSexeSurRepresente &&
    indicateurUnSexeSurRepresente !== indicateurDeuxTroisSexeSurRepresente
  ) {
    return {
      note: barem[0],
      correctionMeasure: true,
      noteEcartTaux,
      noteEcartNombreSalaries,
    }
  }
  if (noteEcartTaux === undefined || noteEcartNombreSalaries === undefined) {
    return {
      note: undefined,
      correctionMeasure: false,
      noteEcartTaux,
      noteEcartNombreSalaries,
    }
  }

  const note = Math.max(noteEcartTaux, noteEcartNombreSalaries)
  return {
    note,
    correctionMeasure: false,
    noteEcartTaux,
    noteEcartNombreSalaries,
  }
}

/////////
// ALL //
/////////

export default function calculIndicateurDeuxTrois(state: AppState) {
  const { totalNombreSalariesHomme: totalNombreSalariesHommes, totalNombreSalariesFemme: totalNombreSalariesFemmes } =
    totalNombreSalaries(state.effectif.nombreSalaries)

  const plusPetitNombreSalaries = calculPlusPetitNombreSalaries(totalNombreSalariesHommes, totalNombreSalariesFemmes)

  const effectifsIndicateurCalculable = calculEffectifsIndicateurCalculable(
    totalNombreSalariesHommes,
    totalNombreSalariesFemmes,
  )

  const indicateurCalculable = calculIndicateurCalculable(
    state.indicateurDeuxTrois.presenceAugmentationPromotion,
    effectifsIndicateurCalculable,
  )

  const tauxAugmentationPromotionHommes = calculTaux(
    state.indicateurDeuxTrois.nombreAugmentationPromotionHommes,
    totalNombreSalariesHommes,
  )
  const tauxAugmentationPromotionFemmes = calculTaux(
    state.indicateurDeuxTrois.nombreAugmentationPromotionFemmes,
    totalNombreSalariesFemmes,
  )

  // // IEA
  const ecartTauxAugmentationPromotion = calculEcartTauxAugmentationPromotion(
    tauxAugmentationPromotionFemmes,
    tauxAugmentationPromotionHommes,
  )

  const indicateurEcartAugmentationPromotion = calculIndicateurEcartAugmentationPromotion(
    indicateurCalculable,
    ecartTauxAugmentationPromotion,
  )

  const indicateurEcartAugmentationPromotionAbsolute = calculIndicateurEcartAugmentationPromotionAbsolute(
    indicateurEcartAugmentationPromotion,
  )

  const indicateurSexeSurRepresente = calculIndicateurSexeSurRepresente(indicateurEcartAugmentationPromotion)

  // // Ecart en nombre équivalent de salariés
  const indicateurEcartNombreEquivalentSalaries = calculIndicateurEcartNombreEquivalentSalaries(
    indicateurEcartAugmentationPromotionAbsolute,
    totalNombreSalariesHommes,
    totalNombreSalariesFemmes,
  )

  // Mesures correction indicateur 1
  const { indicateurSexeSurRepresente: indicateurUnSexeSurRepresente, noteIndicateurUn } = calculIndicateurUn(state)

  // // NOTE
  const {
    note: noteIndicateurDeuxTrois,
    correctionMeasure,
    noteEcartTaux,
    noteEcartNombreSalaries,
  } = calculNote(
    indicateurEcartAugmentationPromotionAbsolute,
    indicateurEcartNombreEquivalentSalaries,
    noteIndicateurUn,
    indicateurUnSexeSurRepresente,
    indicateurSexeSurRepresente,
  )

  return {
    effectifsIndicateurCalculable,
    indicateurCalculable,
    indicateurEcartAugmentationPromotion: indicateurEcartAugmentationPromotionAbsolute,
    indicateurEcartNombreEquivalentSalaries,
    indicateurSexeSurRepresente,
    noteEcartTaux,
    noteEcartNombreSalaries,
    noteIndicateurDeuxTrois,
    correctionMeasure,
    tauxAugmentationPromotionHommes,
    tauxAugmentationPromotionFemmes,
    plusPetitNombreSalaries,
  }
}
