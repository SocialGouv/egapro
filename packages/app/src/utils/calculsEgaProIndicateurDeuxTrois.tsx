import { AppState } from "../globals.d";

import { roundDecimal } from "./helpers";
import totalNombreSalaries from "./totalNombreSalaries";

import {
  calculEcartTauxAugmentation,
  calculIndicateurEcartAugmentation,
  calculIndicateurSexeSurRepresente,
  calculIndicateurEcartAugmentationAbsolute
} from "../utils/calculsEgaProIndicateurDeux";

const barem = [35, 35, 35, 25, 25, 25, 15, 15, 15, 15, 15, 0];

///////////////////////
// INDICATEUR 2 ET 3 //
///////////////////////

// // ETP
export const calculEcartTauxAugmentationPromotion = calculEcartTauxAugmentation;

// // IEP
export const calculIndicateurEcartAugmentationPromotion = calculIndicateurEcartAugmentation;

export const calculIndicateurEcartAugmentationPromotionAbsolute = calculIndicateurEcartAugmentationAbsolute;

// // IC
export const calculIndicateurCalculable = (
  totalNombreSalariesHommes: number | undefined,
  totalNombreSalariesFemmes: number | undefined
): boolean => {
  return (
    totalNombreSalariesHommes !== undefined &&
    totalNombreSalariesFemmes !== undefined &&
    totalNombreSalariesHommes >= 5 &&
    totalNombreSalariesFemmes >= 5
  );
};

export const calculTaux = (
  nombreSalaries: number | undefined,
  totalNombreSalaries: number | undefined
): number | undefined =>
  nombreSalaries !== undefined &&
  totalNombreSalaries !== undefined &&
  totalNombreSalaries > 0
    ? nombreSalaries / totalNombreSalaries
    : undefined;

export const calculIndicateurEcartNombreEquivalentSalaries = (
  indicateurEcartAugmentationPromotionAbsolute: number | undefined,
  totalNombreSalariesHommes: number | undefined,
  totalNombreSalariesFemmes: number | undefined
): number | undefined => {
  return indicateurEcartAugmentationPromotionAbsolute !== undefined &&
    totalNombreSalariesHommes !== undefined &&
    totalNombreSalariesFemmes !== undefined
    ? roundDecimal(
        (indicateurEcartAugmentationPromotionAbsolute *
          Math.min(totalNombreSalariesHommes, totalNombreSalariesFemmes)) /
          100,
        1
      )
    : undefined;
};

// // NOTE

export const calculBarem = (indicateur: number): number => {
  return barem[
    Math.min(
      barem.length - 1,
      Math.ceil(Math.max(0, roundDecimal(indicateur, 1)))
    )
  ];
};

export const calculNote = (
  ecartTaux: number | undefined,
  ecartNombreSalaries: number | undefined
): number | undefined => {
  if (ecartTaux === undefined || ecartNombreSalaries === undefined) {
    return undefined;
  }
  const noteEcartTaux = calculBarem(ecartTaux);
  const noteEcartNombreSalaries = calculBarem(ecartNombreSalaries);

  return Math.max(noteEcartTaux, noteEcartNombreSalaries);
};

/////////
// ALL //
/////////

export default function calculIndicateurDeuxTrois(state: AppState) {
  const {
    totalNombreSalariesHomme: totalNombreSalariesHommes,
    totalNombreSalariesFemme: totalNombreSalariesFemmes
  } = totalNombreSalaries(state.effectif.nombreSalaries);

  const indicateurCalculable = calculIndicateurCalculable(
    totalNombreSalariesHommes,
    totalNombreSalariesFemmes
  );

  const tauxAugmentationPromotionHommes = calculTaux(
    state.indicateurDeuxTrois.nombreAugmentationPromotionHommes,
    totalNombreSalariesHommes
  );
  const tauxAugmentationPromotionFemmes = calculTaux(
    state.indicateurDeuxTrois.nombreAugmentationPromotionFemmes,
    totalNombreSalariesFemmes
  );

  // // IEA
  const ecartTauxAugmentationPromotion = calculEcartTauxAugmentationPromotion(
    tauxAugmentationPromotionFemmes,
    tauxAugmentationPromotionHommes
  );

  const indicateurEcartAugmentationPromotion = calculIndicateurEcartAugmentationPromotion(
    indicateurCalculable,
    ecartTauxAugmentationPromotion
  );

  const indicateurEcartAugmentationPromotionAbsolute = calculIndicateurEcartAugmentationPromotionAbsolute(
    indicateurEcartAugmentationPromotion
  );

  const indicateurSexeSurRepresente = calculIndicateurSexeSurRepresente(
    indicateurEcartAugmentationPromotion
  );

  // // Ecart en nombre équivalent de salariés
  const indicateurEcartNombreEquivalentSalaries = calculIndicateurEcartNombreEquivalentSalaries(
    indicateurEcartAugmentationPromotionAbsolute,
    totalNombreSalariesHommes,
    totalNombreSalariesFemmes
  );

  // // NOTE
  const noteIndicateurDeuxTrois = calculNote(
    indicateurEcartAugmentationPromotionAbsolute,
    indicateurEcartNombreEquivalentSalaries
  );

  return {
    indicateurCalculable,
    indicateurEcartAugmentationPromotion: indicateurEcartAugmentationPromotionAbsolute,
    indicateurEcartNombreEquivalentSalaries,
    indicateurSexeSurRepresente,
    noteIndicateurDeuxTrois
  };
}
