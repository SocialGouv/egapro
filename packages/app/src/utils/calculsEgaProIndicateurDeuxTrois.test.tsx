import stateComplete from "../__fixtures__/stateComplete";
import { calculBarem, calculNote } from "./calculsEgaProIndicateurDeuxTrois";
import calculIndicateurDeuxTrois from "./calculsEgaProIndicateurDeuxTrois";
import { AppState } from "../globals";

///////////////////////
// INDICATEUR 2 et 3 //
///////////////////////

describe("calculBarem", () => {
  test("test barem", () => {
    expect(calculBarem(0)).toEqual(35);
    expect(calculBarem(2)).toEqual(35);
    expect(calculBarem(2.1)).toEqual(25);
    expect(calculBarem(5)).toEqual(25);
    expect(calculBarem(5.1)).toEqual(15);
    expect(calculBarem(10)).toEqual(15);
    expect(calculBarem(10.1)).toEqual(0);
    expect(calculBarem(100)).toEqual(0);
  });
});

describe("calculNote", () => {
  test("cant calcul note", () => {
    expect(calculNote(undefined, undefined)).toEqual(undefined);
    expect(calculNote(1, undefined)).toEqual(undefined);
    expect(calculNote(undefined, 1)).toEqual(undefined);
  });

  test("test max of both notes", () => {
    expect(calculNote(15, 0)).toEqual(35);
    expect(calculNote(0, 15)).toEqual(35);
  });
});

describe("test calculIndicateurDeuxTrois", () => {
  const state: AppState = {
    ...stateComplete,
    effectif: {
      formValidated: "Valid",
      nombreSalaries: [
        {
          categorieSocioPro: 0,
          tranchesAges: [
            {
              nombreSalariesFemmes: 0,
              nombreSalariesHommes: 2,
              trancheAge: 0
            },
            {
              nombreSalariesFemmes: 0,
              nombreSalariesHommes: 1,
              trancheAge: 1
            },
            {
              nombreSalariesFemmes: 0,
              nombreSalariesHommes: 0,
              trancheAge: 2
            },
            {
              nombreSalariesFemmes: 0,
              nombreSalariesHommes: 1,
              trancheAge: 3
            }
          ]
        },
        {
          categorieSocioPro: 1,
          tranchesAges: [
            {
              nombreSalariesFemmes: 9,
              nombreSalariesHommes: 4,
              trancheAge: 0
            },
            {
              nombreSalariesFemmes: 8,
              nombreSalariesHommes: 3,
              trancheAge: 1
            },
            {
              nombreSalariesFemmes: 4,
              nombreSalariesHommes: 3,
              trancheAge: 2
            },
            {
              nombreSalariesFemmes: 2,
              nombreSalariesHommes: 1,
              trancheAge: 3
            }
          ]
        },
        {
          categorieSocioPro: 2,
          tranchesAges: [
            {
              nombreSalariesFemmes: 2,
              nombreSalariesHommes: 1,
              trancheAge: 0
            },
            {
              nombreSalariesFemmes: 2,
              nombreSalariesHommes: 2,
              trancheAge: 1
            },
            {
              nombreSalariesFemmes: 3,
              nombreSalariesHommes: 6,
              trancheAge: 2
            },
            {
              nombreSalariesFemmes: 0,
              nombreSalariesHommes: 2,
              trancheAge: 3
            }
          ]
        },
        {
          categorieSocioPro: 3,
          tranchesAges: [
            {
              nombreSalariesFemmes: 0,
              nombreSalariesHommes: 1,
              trancheAge: 0
            },
            {
              nombreSalariesFemmes: 0,
              nombreSalariesHommes: 1,
              trancheAge: 1
            },
            {
              nombreSalariesFemmes: 0,
              nombreSalariesHommes: 0,
              trancheAge: 2
            },
            {
              nombreSalariesFemmes: 0,
              nombreSalariesHommes: 0,
              trancheAge: 3
            }
          ]
        }
      ]
    },
    indicateurDeuxTrois: {
      formValidated: "Valid",
      memePeriodeReference: true,
      nombreAugmentationPromotionFemmes: 2,
      nombreAugmentationPromotionHommes: 5,
      presenceAugmentationPromotion: true,
      periodeReferenceDebut: "",
      periodeReferenceFin: ""
    }
  };
  test("test data", () => {
    expect(calculIndicateurDeuxTrois(state)).toMatchSnapshot();
  });
});
