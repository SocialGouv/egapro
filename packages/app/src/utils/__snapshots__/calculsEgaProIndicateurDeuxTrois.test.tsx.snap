// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test calculIndicateurDeuxTrois test data 1`] = `
Object {
  "indicateurCalculable": true,
  "indicateurEcartAugmentationPromotion": 11.1905,
  "indicateurEcartNombreEquivalentSalaries": 3.1,
  "indicateurSexeSurRepresente": "hommes",
  "noteIndicateurDeuxTrois": 25,
}
`;
